server/src/main/java/org/opensearch/common/StreamLimiter.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.store.RateLimiter;
+
+import java.io.IOException;
+import java.util.function.Supplier;
+
+/**
+ * The stream limiter that limits the transfer of bytes
+ */
+public class StreamLimiter {
+
+    private final Supplier<RateLimiter> rateLimiterSupplier;
+
+    private final StreamLimiter.Listener listener;
+
+    private int bytesSinceLastRateLimit;
+
+    private static final Logger logger = LogManager.getLogger(StreamLimiter.class);
+
+    public StreamLimiter(Supplier<RateLimiter> rateLimiterSupplier, Listener listener) {
+        this.rateLimiterSupplier = rateLimiterSupplier;
+        this.listener = listener;
+    }
+
+    public void maybePause(int bytes) throws IOException {
+        bytesSinceLastRateLimit += bytes;
+        final RateLimiter rateLimiter = rateLimiterSupplier.get();
+        if (rateLimiter != null) {
+            if (bytesSinceLastRateLimit >= rateLimiter.getMinPauseCheckBytes()) {
+                long pause = rateLimiter.pause(bytesSinceLastRateLimit);
```

Comment:
Should we upgrade the logic here to pause only for the additional bytes that we are receiving now and not the `bytesSinceLastRateLimit`? Haven't we processed the `bytesSinceLastRateLimit` already (the value that is before we add the new `bytes` to it). For snapshot and restore, this would have probably made sense since the snapshot or restore would happen in one go and not a function of indexing. With remote store, the `bytesSinceLastRateLimit` can spread over 1s or last 1 hr depending on the traffic the cluster is subjected to. We might be throttling unnecessary even when there should not be any.

While I know this has worked earlier for snapshot and restore, the previous logic seems to be an overkill one for remote store.

Reply:
I did take a look at the pause method and it seems to considering the overall bytes transferred during the 5 ms duration. Taking back the previous comment, the logic seems to be fine.